File: app/src/main/java/info/nightscout/androidaps/plugins/ConstraintsObjectives/objectives/Objective.java
package info.nightscout.androidaps.plugins.ConstraintsObjectives.objectives;

import android.support.annotation.StringRes;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.R;
import info.nightscout.utils.SP;

public abstract class Objective {

    private int number;
    @StringRes
    private int objective;
    @StringRes
    private int gate;
    private Date startedOn;
    private Date accomplishedOn;
    private List<Task> tasks = new ArrayList<>();

    public Objective(int number, @StringRes int objective, @StringRes int gate) {
        this.number = number;
        this.objective = objective;
        this.gate = gate;
        startedOn = new Date(SP.getLong("Objectives" + number + "started", 0L));
        if (startedOn.getTime() == 0L) startedOn = null;
        accomplishedOn = new Date(SP.getLong("Objectives" + number + "accomplished", 0L));
        if (accomplishedOn.getTime() == 0L) accomplishedOn = null;
        setupTasks(tasks);
        for (Task task : tasks) task.objective = this;
    }

    public boolean isAccomplished() {
        boolean completed = true;
        for (Task task : tasks) {
            if (!task.shouldBeIgnored() && !task.isCompleted()) {
                completed = false;
                break;
            }
        }
        return startedOn != null && (accomplishedOn != null || completed);
    }

    public boolean isStarted() {
        return startedOn != null;
    }

    public Date getStartedOn() {
        return startedOn;
    }

    public int getObjective() {
        return objective;
    }

    public int getGate() {
        return gate;
    }

    public void setStartedOn(Date startedOn) {
        this.startedOn = startedOn;
        SP.putLong("Objectives" + number + "started", startedOn == null ? 0 : startedOn.getTime());
    }

    public void setAccomplishedOn(Date accomplishedOn) {
        this.accomplishedOn = accomplishedOn;
        SP.putLong("Objectives" + number + "accomplished", accomplishedOn == null ? 0 : accomplishedOn.getTime());
    }

    protected void setupTasks(List<Task> tasks) {

    }

    public List<Task> getTasks() {
        return tasks;
    }

    public abstract class Task {
        @StringRes
        private int task;
        private Objective objective;

        public Task(@StringRes int task) {
            this.task = task;
        }

        public int getTask() {
            return task;
        }

        protected Objective getObjective() {
            return objective;
        }

        public abstract boolean isCompleted();

        public String getProgress() {
            return MainApp.gs(isCompleted() ? R.string.completed_well_done : R.string.not_completed_yet);
        }

        public boolean shouldBeIgnored() {
            return false;
        }
    }

    public class MinimumDurationTask extends Task {

        private long minimumDuration;

        public MinimumDurationTask(long minimumDuration) {
            super(R.string.time_leftover);
            this.minimumDuration = minimumDuration;
        }

        @Override
        public boolean isCompleted() {
            return getObjective().isStarted() && System.currentTimeMillis() - getObjective().getStartedOn().getTime() >= minimumDuration;
        }

        @Override
        public String getProgress() {
            long timeLeftover = minimumDuration - (System.currentTimeMillis() - getObjective().getStartedOn().getTime());
            int days = (int) (minimumDuration / (24L * 60L * 60L * 1000L));
            int hours = (int) (minimumDuration / (60L * 60L * 1000L));
            int minutes = (int) (minimumDuration / (60L * 1000L));
            if (days > 0) return MainApp.gq(R.plurals.objective_days, days, days);
            else if (hours > 0) return MainApp.gq(R.plurals.objective_hours, hours, hours);
            else if (minutes > 0) return MainApp.gq(R.plurals.objective_minutes, minutes, minutes);
            else if (timeLeftover > 0) return MainApp.gq(R.plurals.objective_minutes, 1, 1);
            else return MainApp.gs(R.string.time_none);
        }
    }

}

File: plugins/constraints/src/commonMain/kotlin/app/aaps/plugins/constraints/objectives/objectives/Objective.kt
package app.aaps.plugins.constraints.objectives.objectives

import app.aaps.core.keys.interfaces.TextRef
import app.aaps.plugins.constraints.ConstraintsStrings
import app.aaps.core.data.time.T
import app.aaps.core.interfaces.resources.TextResolver
import app.aaps.core.interfaces.utils.DateUtil
import app.aaps.core.keys.interfaces.Preferences
import app.aaps.plugins.constraints.objectives.keys.ObjectivesBooleanComposedKey
import app.aaps.plugins.constraints.objectives.keys.ObjectivesLongComposedKey
import kotlinx.coroutines.Runnable
import kotlin.math.floor
import kotlin.time.Clock

abstract class Objective(
    val preferences: Preferences,
    val rh: TextResolver,
    val dateUtil: DateUtil,
    /** Renders "3 d" and friends. A parameter because Android says it with plurals; see [DurationText]. */
    val durationText: DurationText,
    private val spName: String,
    val objective: TextRef,
    val gate: TextRef
) {

    var startedOn: Long = 0
        get() = preferences.get(ObjectivesLongComposedKey.Started, spName)
        set(value) {
            field = value
            preferences.put(ObjectivesLongComposedKey.Started, spName, value = value)
        }
    var accomplishedOn: Long = 0
        get() {
            var value = preferences.get(ObjectivesLongComposedKey.Accomplished, spName)
            if (value - dateUtil.now() > T.hours(3).msecs() || startedOn - dateUtil.now() > T.hours(3).msecs()) { // more than 3 hours in the future
                startedOn = 0
                accomplishedOn = 0
                value = 0
            }
            return value
        }
        set(value) {
            field = value
            preferences.put(ObjectivesLongComposedKey.Accomplished, spName, value = value)
        }

    var tasks: MutableList<Task> = ArrayList()

    suspend fun isCompleted(): Boolean {
        for (task in tasks) {
            if (!task.shouldBeIgnored() && !task.isCompleted()) return false
        }
        return true
    }

    suspend fun isCompleted(trueTime: Long): Boolean {
        for (task in tasks) {
            if (!task.shouldBeIgnored() && !task.isCompleted(trueTime)) return false
        }
        return true
    }

    val isAccomplished: Boolean
        get() = accomplishedOn != 0L && accomplishedOn < dateUtil.now()
    val isStarted: Boolean
        get() = startedOn != 0L

    abstract inner class Task(var objective: Objective, val task: TextRef) {

        var hints = ArrayList<Hint>()
        var learned = ArrayList<Learned>()

        abstract suspend fun isCompleted(): Boolean

        open suspend fun isCompleted(trueTime: Long): Boolean = isCompleted()

        open suspend fun progress(): String =
            rh.gs(if (isCompleted()) ConstraintsStrings.completed_well_done else ConstraintsStrings.not_completed_yet)

        fun hint(hint: Hint): Task {
            hints.add(hint)
            return this
        }

        fun learned(learned: Learned): Task {
            this.learned.add(learned)
            return this
        }

        open fun shouldBeIgnored(): Boolean = false
    }

    inner class MinimumDurationTask internal constructor(objective: Objective, private val minimumDuration: Long) : Task(objective, ConstraintsStrings.time_elapsed) {

        override suspend fun isCompleted(): Boolean =
            objective.isStarted && Clock.System.now().toEpochMilliseconds() - objective.startedOn >= minimumDuration

        override suspend fun isCompleted(trueTime: Long): Boolean {
            return objective.isStarted && trueTime - objective.startedOn >= minimumDuration
        }

        override suspend fun progress(): String =
            (getDurationText(Clock.System.now().toEpochMilliseconds() - objective.startedOn)
                + " / " + getDurationText(minimumDuration))

        private fun getDurationText(duration: Long): String = durationText.format(duration)
    }

    inner class UITask internal constructor(objective: Objective, task: TextRef, private val spIdentifier: String, val code: (task: UITask, callback: Runnable, showMessage: (String) -> Unit) -> Unit) : Task(objective, task) {

        var answered: Boolean = false
            set(value) {
                field = value
                preferences.put(ObjectivesBooleanComposedKey.AnsweredUi, spIdentifier, value = value)
            }

        init {
            answered = preferences.get(ObjectivesBooleanComposedKey.AnsweredUi, spIdentifier)
        }

        override suspend fun isCompleted(): Boolean = answered
    }

    inner class ExamTask internal constructor(objective: Objective, task: TextRef, val question: TextRef, private val spIdentifier: String) : Task(objective, task) {

        var options = ArrayList<Option>()
        var answered: Boolean = false
            set(value) {
                field = value
                preferences.put(ObjectivesBooleanComposedKey.AnsweredExam, spIdentifier, value = value)
            }
        var disabledTo: Long = 0
            set(value) {
                field = value
                preferences.put(ObjectivesLongComposedKey.DisabledTo, spIdentifier, value = value)
            }

        init {
            answered = preferences.get(ObjectivesBooleanComposedKey.AnsweredExam, spIdentifier)
            disabledTo = preferences.get(ObjectivesLongComposedKey.DisabledTo, spIdentifier)
        }

        override suspend fun isCompleted(): Boolean = answered

        fun isEnabledAnswer(): Boolean = disabledTo < dateUtil.now()

        fun option(option: Option): ExamTask {
            options.add(option)
            return this
        }
    }

    class Option internal constructor(var option: TextRef, var isCorrect: Boolean)

    class Hint internal constructor(var hint: TextRef)

    class Learned internal constructor(var learned: TextRef)
}